spec/models/site_status_spec.rb

```diff
@@ -117,4 +117,60 @@
       it { should_not have_vacancies }
     end
   end
+
+  describe "#update" do
+    all_vac_statuses = %w[no_vacancies part_time_vacancies full_time_vacancies both_full_time_and_part_time_vacancies]
+    specs = [
+      {
+        study_mode: :with_course_study_mode_as_nil,
+        valid_states: %w[no_vacancies]
+      },
+      {
+        study_mode: :with_course_study_mode_as_full_time,
+        valid_states: %w[no_vacancies full_time_vacancies]
+      },
+      {
+        study_mode: :with_course_study_mode_as_part_time,
+        valid_states: %w[no_vacancies part_time_vacancies]
+      },
+      {
+        study_mode: :with_course_study_mode_as_full_time_or_part_time,
+        valid_states: all_vac_statuses
+      },
+    ].freeze
+
+    specs.each do |spec|
```

Comment:
In Rails-speak what you're testing here is that an object is valid with certain combinations of values. I think these tests would be clearer if focused on that and didn't worry so much about making them DRY. This would be better, with 

```
  describe 'vac_status' do
    context 'full time course' do
      let(:course) { build(:course, study_mode: :full_time) }

      context 'vac_status set to full_time' do
        subject { create(:site_status, vac_status: :full_time_vacancies, course: course) }
        it { should be_valid }
      end

      context 'vac_status set to part_time' do
        subject { create(:site_status, :allow_invalid, vac_status: :part_time_vacancies, course: course) }
        it { should_not be_valid }
      end
    end
```

If we really want to DRY them up, I think a custom matcher would be better. It breaks the one-expectation per test guideline, but it is still readable:

```
    describe 'vac_status' do
      RSpec::Matchers.define :be_valid_vac_status_for_course do |course|
        match do |vac_status|
          site_status = create(:site_status,
                               :allow_invalid,
                               vac_status: vac_status,
                               course: course)
          site_status.valid?
        end
      end

      context 'full time course' do
        let(:full_time_course) { create(:course, study_mode: :full_time) }

        it 'only allows full_time or no vacancies' do
          expect(:no_vacancies).to be_valid_vac_status_for_course(full_time_course)
          expect(:full_time_vacancies).to be_valid_vac_status_for_course(full_time_course)
          expect(:part_time_vacancies).not_to be_valid_vac_status_for_course(full_time_course)
          expect(:both_full_time_and_part_time_vacancies).not_to be_valid_vac_status_for_course(full_time_course)
        end
      end
    end
```

Here is the `allow_invalid` trait for `site_status`. I was a little surprised we need it, but I guess it makes sense for factories to insist that the objects they create are valid, and it might make sense to bypass the factory to create invalid objects in model tests.

```
    trait :allow_invalid do
      to_create { |instance| instance.save }
    end
```